feature/VMO-7721/SupportPhp81

AlteaDown opened the pull request:
Added PHP 8.1 support, as in upgrading the Viamo platform to PHP 8.1

I had identified this project does not actually support PHP 8.1, even though composer says it does, so I made it compatible with PHP 8.1.

I also added types, where supported by PHP 8.1 and where I was able to add them in bulk by IDE, especially where inheritance/abstraction/interfaces were involved, so that issues were immediately apparent. 

We should release this as a major revision, so that we can drop php <8 support and laravel <8 support.

AlteaDown committed:
feature/VMO-7721/SupportPhp81 - Added PHP 8.1 support, as in upgrading the Viamo platform to PHP 8.1, I had identified this project does not actually support PHP 8.1, even though composer says it does. I also added types, where supported by PHP 8.1, and especially where inheritance was involved, so that issues were immediately apparent. We should release this as a major revision, so that we can drop php <8 support and laravel <8 support.

AlteaDown committed:
feature/VMO-7721/SupportPhp81 - Removed PHP 7.4 support.

seifertk commented on dist/BaseExpressionParser.php:
```diff
@@ -290,11 +290,14 @@ private function peg_buildException($message, $expected, $pos) {
-        return $ex; 
-      }
-    private function peg_f1($id, $ex) { return call_user_func($this->_location); }
+    private function peg_f0($id, $ex) {
```
FYI this is a generated file so I wouldn't expect these changes to persist.

lakhbawa replied:
approved

AlteaDown committed:
feature/VMO-7721/SupportPhp81 - fixed whitespace changes. I forgot the codestyle is different in this project, and didn't notice I had the "Reformat Code" option enabled for by commit.

AlteaDown committed:
Merge branch 'master' of github.com:FLOIP/expression-parser into feature/VMO-7721/SupportPhp81

# Conflicts:
#	src/php/Evaluator/LogicNodeEvaluator.php

AlteaDown committed:
feature/VMO-7721/SupportPhp81 - fixed whitespace changes. I forgot the codestyle is different in this project, and didn't notice I had the "Reformat Code" option enabled for by commit. I added support for Rector, and ran it against the project, to speed up the migration to PHP 8.1. I also found additional cases I had to handle, and made changes as necessary. Added Types wherever possible. Found an issue with Text->substitute, consulted Kyle, and then implemented the expected behaviour, as the issue became apparent when making these changes.

feature/VMO-7721/SupportPhp81 - fixed whitespace changes. I forgot the codestyle is different in this project, and didn't notice I had the "Reformat Code" option enabled for by commit. I added support for Rector, and ran it against the project, to speed up the migration to PHP 8.1. I also found additional cases I had to handle, and made changes as necessary. Added Types wherever possible. Found an issue with Text->substitute, consulted Kyle, and then implemented the expected behaviour, as the issue became apparent when making these changes.

AlteaDown committed:
feature/VMO-7721/SupportPhp81 - Second Rector Run.

AlteaDown committed:
feature/VMO-7721/SupportPhp81 - Unit Tests are now passing.

lakhbawa commented:
approved

AlteaDown committed:
feature/VMO-7721/SupportPhp81 - Reverted changes to some generated files.
